marshal/marshal.go

```diff
@@ -121,27 +121,53 @@ func (m *Marshaler) Partitions(topicName string) int {
 	return m.cluster.getTopicPartitions(topicName)
 }
 
-// Terminate is called when we're done with the marshaler and want to shut down.
-func (m *Marshaler) Terminate() {
+// terminateAndCleanup terminates the marshal, with the option of removing
+// the marshaler's reference from its associated cluster.
+func (m *Marshaler) terminateAndCleanup(remove bool) {
 	if !atomic.CompareAndSwapInt32(m.quit, 0, 1) {
 		return
 	}
 
 	m.lock.Lock()
 	defer m.lock.Unlock()
-
 	// Now terminate all of the consumers. In this codepath we do a no-release termination
 	// because that is usually correct in production. If someone actually wants to release
 	// they need to terminate the consumers manually.
 	for _, cn := range m.consumers {
-		cn.Terminate(false)
+		cn.terminateAndCleanup(false, false)
 	}
 	m.consumers = nil
 
 	// If we own the cluster, terminate it.
 	if m.ownsCluster {
 		m.cluster.Terminate()
 	}
+
+	// Remove this marshal from its cluster. Doing so is recommended
+	// if the cluster doesn't remove the terminated marshal itself (by setting its
+	// list of marshals to nil or filtering them).
+	if remove {
+		var wg sync.WaitGroup
+		wg.Add(1)
```

Comment:
Same comments here as above. I think moving this up to the Cluster object might be better.